fix(resourcemanager): append custom `user-agent` headers

archerzz opened the pull request:
Right now, the mgmt telemetry policy will override `user-agent` header to either default header value, or `UserAgentOverride` which is set in the generated codes of each SDK. Then the custom `user-agent` headers are all lost. That will prevent 3rd parties (like powershell) to do their analytics.

This commit will:
1. change the telemetry policy behavior to append custom headers after the official value, which is also a best practice in the industry
2. add unit tests

fix #26406

# All SDK Contribution checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.
- [ ] **Please open PR in `Draft` mode if it is:**
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.
- [ ] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md).**
- [ ] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#general-guidelines)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. (Track 2 only)
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. Please double check nuget.org current release version.

## Additional management plane SDK specific contribution checklist: 
Note: Only applies to `Microsoft.Azure.Management.[RP]` or `Azure.ResourceManager.[RP]`
 
- [ ] Include updated [management metadata](https://github.com/Azure/azure-sdk-for-net/tree/main/eng/mgmt/mgmtmetadata).
- [ ] Update AzureRP.props to add/remove version info to maintain up to date API versions.

### Management plane SDK Troubleshooting
- If this is very first SDK for a services and you are adding new service folders directly under /SDK, please add `new service` label and/or contact assigned reviewer.
- If the check fails at the `Verify Code Generation` step, please ensure:
	- Do not modify any code in generated folders.
	- Do not selectively include/remove generated files in the PR.
	- Do use `generate.ps1/cmd` to generate this PR instead of calling `autorest` directly.
	Please pay attention to the @microsoft.csharp version output after running `generate.ps1`. If it is lower than current released version (2.3.82), please run it again as it should pull down the latest version.
	
	**Note: We have recently updated the PSH module called by `generate.ps1` to emit additional data. This would help reduce/eliminate the Code Verification check error. Please run following command**:

	    `dotnet msbuild eng/mgmt.proj /t:Util /p:UtilityName=InstallPsModules`

### Old outstanding PR cleanup
 Please note:
	If PRs (including draft) has been out for more than 60 days and there are no responses from our query or followups, they will be closed to maintain a concise list for our reviewers.

archerzz committed:
fix(resourcemanager): append custom `user-agent` headers

Right now, the mgmt telemetry policy will override `user-agent` header to either default header value, or `UserAgentOverride` which is set in the generated codes of each SDK. Then the custom `user-agent` headers are all lost. That will prevent 3rd parties (like powershell) to do their analytics.

This commit will:
1. change the telemetry policy behavior to append custom headers after the official value, which is also a best practice in the industry
2. add unit tests

fix #26406

ArthurMa1978 commented on sdk/resourcemanager/Azure.ResourceManager/src/Internal/MgmtTelemetryPolicy.cs:
```diff
@@ -19,6 +18,11 @@ public MgmtTelemetryPolicy(object source, ClientOptions options)
         public override void OnSendingRequest(HttpMessage message)
         {
             var header = message.TryGetProperty("UserAgentOverride", out var userAgent) ? userAgent as string : _defaultHeader;
+            if (message.Request.Headers.TryGetValues(HttpHeader.Names.UserAgent, out var customHeaders))
```
A little bit confused, if it's 'override' then why need to append? Or this 'override' only override our default user agent string?

archerzz replied:
I think the logic has changed here. Previously we do override the `user-agent`. But now, we cannot override.

However, `UserAgentOverride` is set in the generated SDK codes, like below:
```csharp
            request.Uri = uri;
            request.Headers.Add("Accept", "application/json");
            message.SetProperty("UserAgentOverride", _userAgent);
```

We can consider to change the name of the property.

archerzz replied:
renamed to `SDKUserAgent`

ArthurMa1978 reviewed (approved):
Please check my comment

archerzz committed:
change according to review comments

1. rename `UserAgentOverride` to `SDKUserAgent`, since now we allow to append customer headers
2. change to use `StringBuilder` to join headers
3. add benchmark test

archerzz committed:
remove iteration setup since it is not suitable for microbenchmark

archerzz commented:
After reading this [article](https://dotnetcoretutorials.com/2020/02/06/performance-of-string-concatenation-in-c/), I've decided to use `StringBuilder` to concatenate user agents.

Here are the benchmarks:

### Original (e.g. not handling custom headers)

|                     Method | SDKUserAgent | CustomHeaders |     Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------------- |------------------ |-------------- |---------:|----------:|----------:|-------:|------:|------:|----------:|
| UserAgentOverrideBenchmark |                   |      Array[0] | 7.673 us | 0.1401 us | 0.1310 us | 0.1068 |     - |     - |   3.13 KB |
| UserAgentOverrideBenchmark |                   |      Array[1] | 7.826 us | 0.1559 us | 0.2427 us | 0.1068 |     - |     - |   3.13 KB |
| UserAgentOverrideBenchmark |                   |      Array[2] | 7.878 us | 0.1563 us | 0.1386 us | 0.1068 |     - |     - |   3.22 KB |
| UserAgentOverrideBenchmark |              test |      Array[0] | 7.919 us | 0.0427 us | 0.0356 us | 0.1068 |     - |     - |   3.34 KB |
| UserAgentOverrideBenchmark |              test |      Array[1] | 7.975 us | 0.0754 us | 0.0705 us | 0.1068 |     - |     - |   3.34 KB |
| UserAgentOverrideBenchmark |              test |      Array[2] | 8.096 us | 0.0517 us | 0.0458 us | 0.1221 |     - |     - |   3.43 KB |

### Use `string.Join`

|                     Method | SDKUserAgent | CustomHeaders |     Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------------- |------------------ |-------------- |---------:|----------:|----------:|-------:|------:|------:|----------:|
| UserAgentOverrideBenchmark |                   |      Array[0] | 7.640 us | 0.0750 us | 0.0665 us | 0.1068 |     - |     - |   3.13 KB |
| UserAgentOverrideBenchmark |                   |      Array[1] | 8.146 us | 0.0830 us | 0.0736 us | 0.1221 |     - |     - |   3.49 KB |
| UserAgentOverrideBenchmark |                   |      Array[2] | 8.459 us | 0.1358 us | 0.1270 us | 0.1221 |     - |     - |    3.5 KB |
| UserAgentOverrideBenchmark |              test |      Array[0] | 7.943 us | 0.0542 us | 0.0453 us | 0.1068 |     - |     - |   3.34 KB |
| UserAgentOverrideBenchmark |              test |      Array[1] | 8.367 us | 0.0916 us | 0.0812 us | 0.1221 |     - |     - |   3.55 KB |
| UserAgentOverrideBenchmark |              test |      Array[2] | 8.585 us | 0.1401 us | 0.1242 us | 0.1221 |     - |     - |   3.57 KB |

### Use `StringBuilder`
|                     Method | SDKUserAgent | CustomHeaders |     Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------------- |------------------ |-------------- |---------:|----------:|----------:|-------:|------:|------:|----------:|
| UserAgentOverrideBenchmark |                   |      Array[0] | 7.631 us | 0.0512 us | 0.0428 us | 0.1068 |     - |     - |   3.13 KB |
| UserAgentOverrideBenchmark |                   |      Array[1] | 8.159 us | 0.0692 us | 0.0648 us | 0.1373 |     - |     - |   3.91 KB |
| UserAgentOverrideBenchmark |                   |      Array[2] | 8.286 us | 0.0979 us | 0.0916 us | 0.1373 |     - |     - |   3.92 KB |
| UserAgentOverrideBenchmark |              test |      Array[0] | 7.970 us | 0.0893 us | 0.0835 us | 0.1068 |     - |     - |   3.34 KB |
| UserAgentOverrideBenchmark |              test |      Array[1] | 8.311 us | 0.0858 us | 0.0717 us | 0.1221 |     - |     - |   3.62 KB |
| UserAgentOverrideBenchmark |              test |      Array[2] | 8.485 us | 0.1154 us | 0.1023 us | 0.1221 |     - |     - |   3.63 KB |

I think `StringBuilder` is slightly better than `string.Join` in time, but slightly worse in memory. It's still worthy to use `StringBuilder` since the memory footprint is trivial.

**Note**: The benchmark codes are not perfect. It should move the preparation codes into `[IterationSetup]`, but according to the [BenchmarkDotNet doc](https://benchmarkdotnet.org/articles/features/setup-and-cleanup.html#sample-introsetupcleanupiteration), `[IterationSetup]` is not recommended for microbenchmark. I have tried the following benchmark codes, and the results are weird.

```csharp
        [IterationSetup]
        public void Setup()
        {
            var requestMock = new MockRequest();
            message = new HttpMessage(requestMock, new ResponseClassifier());
            if (!string.IsNullOrEmpty(SDKAgentOverride))
            {
                message.SetProperty("SDKUserAgent", SDKAgentOverride);
            }
            if (CustomHeaders.Length > 0)
            {
                foreach (var header in CustomHeaders)
                {
                    requestMock.Headers.Add(HttpHeader.Names.UserAgent, header);
                }
            }
            policy = new MgmtTelemetryPolicy(message, ClientOptions.Default);
        }

        [Benchmark]
        public void UserAgentOverrideBenchmark()
        {
            policy.OnSendingRequest(message);
        }
```

Although the benchmark for above codes are weird, it still shows that `StringBuilder` is slighter better than `string.Join`.

#### Original
|                     Method | UserAgentOverride | CustomHeaders |     Mean |     Error |    StdDev |   Median | Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------------- |------------------ |-------------- |---------:|----------:|----------:|---------:|------:|------:|------:|----------:|
| UserAgentOverrideBenchmark |                   |      Array[0] | 3.519 us | 0.2422 us | 0.7026 us | 3.450 us |     - |     - |     - |         - |
| UserAgentOverrideBenchmark |                   |      Array[1] | 1.616 us | 0.1693 us | 0.4939 us | 1.550 us |     - |     - |     - |         - |
| UserAgentOverrideBenchmark |                   |      Array[2] | 1.632 us | 0.1577 us | 0.4576 us | 1.550 us |     - |     - |     - |         - |
| UserAgentOverrideBenchmark |              test |      Array[0] | 4.612 us | 0.3265 us | 0.9263 us | 4.200 us |     - |     - |     - |         - |
| UserAgentOverrideBenchmark |              test |      Array[1] | 2.626 us | 0.1953 us | 0.5697 us | 2.450 us |     - |     - |     - |         - |
| UserAgentOverrideBenchmark |              test |      Array[2] | 2.572 us | 0.1788 us | 0.5160 us | 2.500 us |     - |     - |     - |         - |


#### `string.Join`
|                     Method | UserAgentOverride | CustomHeaders |     Mean |     Error |    StdDev |   Median | Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------------- |------------------ |-------------- |---------:|----------:|----------:|---------:|------:|------:|------:|----------:|
| UserAgentOverrideBenchmark |                   |      Array[0] | 3.816 us | 0.2425 us | 0.7035 us | 3.600 us |     - |     - |     - |         - |
| UserAgentOverrideBenchmark |                   |      Array[1] | 5.907 us | 0.2768 us | 0.7761 us | 5.900 us |     - |     - |     - |         - |
| UserAgentOverrideBenchmark |                   |      Array[2] | 6.060 us | 0.3490 us | 1.0070 us | 6.000 us |     - |     - |     - |         - |
| UserAgentOverrideBenchmark |              test |      Array[0] | 4.688 us | 0.2219 us | 0.6330 us | 4.550 us |     - |     - |     - |         - |
| UserAgentOverrideBenchmark |              test |      Array[1] | 6.580 us | 0.2834 us | 0.7992 us | 6.400 us |     - |     - |     - |         - |
| UserAgentOverrideBenchmark |              test |      Array[2] | 6.155 us | 0.2147 us | 0.6159 us | 6.100 us |     - |     - |     - |         - |

### `StringBuilder`
|                     Method | UserAgentOverride | CustomHeaders |     Mean |     Error |    StdDev |   Median | Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------------- |------------------ |-------------- |---------:|----------:|----------:|---------:|------:|------:|------:|----------:|
| UserAgentOverrideBenchmark |                   |      Array[0] | 3.595 us | 0.2349 us | 0.6815 us | 3.400 us |     - |     - |     - |         - |
| UserAgentOverrideBenchmark |                   |      Array[1] | 4.956 us | 0.2451 us | 0.7112 us | 4.850 us |     - |     - |     - |         - |
| UserAgentOverrideBenchmark |                   |      Array[2] | 4.834 us | 0.2069 us | 0.5970 us | 4.600 us |     - |     - |     - |         - |
| UserAgentOverrideBenchmark |              test |      Array[0] | 4.543 us | 0.2391 us | 0.6937 us | 4.400 us |     - |     - |     - |         - |
| UserAgentOverrideBenchmark |              test |      Array[1] | 5.528 us | 0.2350 us | 0.6781 us | 5.400 us |     - |     - |     - |         - |
| UserAgentOverrideBenchmark |              test |      Array[2] | 5.312 us | 0.2281 us | 0.6509 us | 5.200 us |     - |     - |     - |         - |

archerzz committed:
change back to `string.Join` since we favor memory footprint

archerzz commented:
After discussion, we change back to `string.Join` because of the smaller memory footprint, considering the difference in time consumption is trivial.
